File: unity/Assets/_Project/Scripts/Core/Rules/CutThroatRules.cs
#nullable enable
using System;
using System.Collections.Generic;

namespace Pose.Core
{
    /// <summary>
    /// Jamaican Cut-Throat dominoes — every player for themselves, no teams. Standard
    /// double-six tile set, 2–4 players, no boneyard draws: a player who cannot match
    /// either chain end must pass. The round ends either when a player empties their
    /// hand (<see cref="MatchEndReason.Domino"/>) or when every player passes in
    /// succession (<see cref="MatchEndReason.Blocked"/>). Round winner scores the sum
    /// of every other player's remaining pips; a tied block ends in a draw with no
    /// score. Multi-round mechanics ("Six-Love", tied-block replay for 2 points) are
    /// out of scope at this layer — the engine produces a single-round outcome that a
    /// higher tournament/scoring layer wraps. The "Block / Draw (Anglo)" variant in
    /// <c>docs/ARCHITECTURE.md</c> §8.1 is a separate future ruleset; this type
    /// implements the Jamaican Cut-Throat row of that table.
    /// </summary>
    public sealed class CutThroatRules : IRuleEngine
    {
        private readonly byte _maxPip;

        /// <summary>
        /// Creates a Cut-Throat rule engine for a tile set with the given maximum pip
        /// value (6 for standard double-six). The max pip is needed only by the
        /// opening-turn rule to identify the leading double.
        /// </summary>
        public CutThroatRules(byte maxPip = 6)
        {
            _maxPip = maxPip;
        }

        public IReadOnlyList<Move> GetLegalMoves(MatchState state)
        {
            if (state == null)
            {
                throw new ArgumentNullException(nameof(state));
            }

            if (state.IsOver)
            {
                return Array.Empty<Move>();
            }

            PlayerId player = state.CurrentPlayer;
            Hand hand = state.Hands[player];
            List<Move> moves = new();

            if (state.Chain.IsEmpty)
            {
                // Opening turn — only the leading tile (highest double, or highest
                // single tile if no doubles exist anywhere) may be played, by its
                // holder. The Dealer set CurrentPlayer to that holder; here we
                // re-derive the tile so the rule engine remains a pure function of
                // state with no Dealer-private knowledge.
                StartingPlayerRule.Lead lead = StartingPlayerRule.FindLead(
                    state.Players,
                    state.Hands,
                    _maxPip);

                moves.Add(new PlaceMove(player, lead.Tile, ChainEnd.Left));
                return moves;
            }

            byte left = state.Chain.LeftEnd;
            byte right = state.Chain.RightEnd;

            foreach (Tile tile in hand)
            {
                if (tile.Matches(left))
                {
                    moves.Add(new PlaceMove(player, tile, ChainEnd.Left));
                }

                if (tile.Matches(right))
                {
                    moves.Add(new PlaceMove(player, tile, ChainEnd.Right));
                }
            }

            if (moves.Count == 0)
            {
                moves.Add(new PassMove(player));
            }

            return moves;
        }

        public bool IsLegal(MatchState state, Move move)
        {
            if (state == null)
            {
                throw new ArgumentNullException(nameof(state));
            }

            if (move == null)
            {
                throw new ArgumentNullException(nameof(move));
            }

            // Identity check: the move's player must be the current player. A move
            // submitted by anyone else is rejected as a structural violation.
            if (move.Player != state.CurrentPlayer)
            {
                return false;
            }

            IReadOnlyList<Move> legal = GetLegalMoves(state);
            for (int i = 0; i < legal.Count; i++)
            {
                if (MovesEqual(legal[i], move))
                {
                    return true;
                }
            }
            return false;
        }

        public MatchState Apply(MatchState state, Move move)
        {
            if (state == null)
            {
                throw new ArgumentNullException(nameof(state));
            }

            if (move == null)
            {
                throw new ArgumentNullException(nameof(move));
            }

            if (state.IsOver)
            {
                throw new InvalidOperationException(
                    "Cannot apply a move to a finished match.");
            }

            if (!IsLegal(state, move))
            {
                throw new InvalidOperationException($"Illegal move: {move}");
            }

            List<Move> newHistory = new(state.History) { move };

            if (move is PlaceMove pm)
            {
                Hand newHand = state.Hands[pm.Player].Without(pm.Tile);
                Dictionary<PlayerId, Hand> newHands = new(state.Hands)
                {
                    [pm.Player] = newHand,
                };
                Chain newChain = state.Chain.Place(pm.Tile, pm.End);

                bool wonByDomino = newHand.Count == 0;
                int nextIdx = wonByDomino
                    ? state.CurrentPlayerIndex
                    : NextPlayerIndex(state);

                return state.With(
                    currentPlayerIndex: nextIdx,
                    hands: newHands,
                    chain: newChain,
                    turnNumber: state.TurnNumber + 1,
                    consecutivePassCount: 0,
                    history: newHistory,
                    isOver: wonByDomino);
            }

            if (move is PassMove)
            {
                int newPassCount = state.ConsecutivePassCount + 1;
                bool blocked = newPassCount >= state.Players.Count;
                int nextIdx = blocked
                    ? state.CurrentPlayerIndex
                    : NextPlayerIndex(state);

                return state.With(
                    currentPlayerIndex: nextIdx,
                    turnNumber: state.TurnNumber + 1,
                    consecutivePassCount: newPassCount,
                    history: newHistory,
                    isOver: blocked);
            }

            throw new ArgumentException(
                $"Unsupported move type: {move.GetType().Name}",
                nameof(move));
        }

        public MatchOutcome? GetOutcome(MatchState state)
        {
            if (state == null)
            {
                throw new ArgumentNullException(nameof(state));
            }

            if (!state.IsOver)
            {
                return null;
            }

            Dictionary<PlayerId, int> remaining = new(state.Players.Count);
            for (int i = 0; i < state.Players.Count; i++)
            {
                PlayerId p = state.Players[i];
                remaining[p] = state.Hands[p].PipTotal;
            }

            // Check for Domino end: someone has zero tiles.
            PlayerId? dominoWinner = null;
            for (int i = 0; i < state.Players.Count; i++)
            {
                PlayerId p = state.Players[i];
                if (state.Hands[p].Count == 0)
                {
                    dominoWinner = p;
                    break;
                }
            }

            if (dominoWinner.HasValue)
            {
                int score = 0;
                for (int i = 0; i < state.Players.Count; i++)
                {
                    PlayerId p = state.Players[i];
                    if (p != dominoWinner.Value)
                    {
                        score += state.Hands[p].PipTotal;
                    }
                }

                return new MatchOutcome(
                    MatchEndReason.Domino,
                    dominoWinner,
                    state.Partnership.GetTeamOf(dominoWinner.Value),
                    score,
                    remaining);
            }

            // Block end: lowest pip total wins, ties produce a draw with no points.
            PlayerId? blockWinner = null;
            int lowest = int.MaxValue;
            bool tiedAtLowest = false;
            for (int i = 0; i < state.Players.Count; i++)
            {
                PlayerId p = state.Players[i];
                int pips = state.Hands[p].PipTotal;
                if (pips < lowest)
                {
                    lowest = pips;
                    blockWinner = p;
                    tiedAtLowest = false;
                }
                else if (pips == lowest)
                {
                    tiedAtLowest = true;
                }
            }

            if (tiedAtLowest)
            {
                return new MatchOutcome(
                    MatchEndReason.Blocked,
                    winnerId: null,
                    winningTeamId: null,
                    winnerScore: 0,
                    remaining);
            }

            int blockScore = 0;
            for (int i = 0; i < state.Players.Count; i++)
            {
                PlayerId p = state.Players[i];
                if (p != blockWinner!.Value)
                {
                    blockScore += state.Hands[p].PipTotal;
                }
            }

            // blockWinner is guaranteed non-null here: the loop above always
            // runs (a match has >= 2 players) and assigns blockWinner on the
            // first iteration, and the tiedAtLowest early-return is the only
            // path that skips this point.
            return new MatchOutcome(
                MatchEndReason.Blocked,
                blockWinner,
                state.Partnership.GetTeamOf(blockWinner!.Value),
                blockScore,
                remaining);
        }

        private static int NextPlayerIndex(MatchState state) =>
            (state.CurrentPlayerIndex + 1) % state.Players.Count;

        private static bool MovesEqual(Move a, Move b)
        {
            if (a.Player != b.Player)
            {
                return false;
            }

            if (a is PlaceMove ap && b is PlaceMove bp)
            {
                return ap.Tile == bp.Tile && ap.End == bp.End;
            }

            if (a is PassMove && b is PassMove)
            {
                return true;
            }

            return false;
        }
    }
}

File: functions/src/rules/cutThroatRules.ts
import { PlayerId } from './ids';
import { MatchEndReason, MatchOutcome } from './matchOutcome';
import { MatchState } from './matchState';
import { Move, PlaceMove, placeMove, passMove } from './move';
import { findLead } from './startingPlayerRule';

/**
 * Jamaican Cut-Throat rules — every player for themselves, no boneyard. Port of
 * `Pose.Core.CutThroatRules`. A round ends on a domino (empty hand), a block
 * (every player passing in succession), or a resign. The winner scores the sum
 * of the other players' remaining pips; a tied block is a scoreless draw.
 */
export class CutThroatRules {
  private readonly maxPip: number;

  constructor(maxPip = 6) {
    this.maxPip = maxPip;
  }

  getLegalMoves(state: MatchState): Move[] {
    if (state.isOver) {
      return [];
    }

    const player = state.currentPlayer;
    const hand = state.handOf(player);

    if (state.chain.isEmpty) {
      // Opening turn — only the leading tile, by its holder. Re-derived from
      // state so the engine stays a pure function with no dealer-private input.
      const lead = findLead(state.players, state.hands, this.maxPip);
      return [placeMove(player, lead.tile, 'left')];
    }

    const left = state.chain.leftEnd;
    const right = state.chain.rightEnd;
    const moves: Move[] = [];
    for (const tile of hand.tiles) {
      if (tile.matches(left)) {
        moves.push(placeMove(player, tile, 'left'));
      }
      if (tile.matches(right)) {
        moves.push(placeMove(player, tile, 'right'));
      }
    }
    if (moves.length === 0) {
      moves.push(passMove(player));
    }
    return moves;
  }

  isLegal(state: MatchState, move: Move): boolean {
    if (state.isOver) {
      return false;
    }

    // Resign is legal for any participant at any time (disconnect / forfeit).
    if (move.kind === 'resign') {
      return state.players.includes(move.player);
    }

    if (move.player !== state.currentPlayer) {
      return false;
    }

    const legal = this.getLegalMoves(state);
    return legal.some((m) => movesEqual(m, move));
  }

  apply(state: MatchState, move: Move): MatchState {
    if (state.isOver) {
      throw new Error('Cannot apply a move to a finished match.');
    }
    if (!this.isLegal(state, move)) {
      throw new Error(`Illegal move: ${move.kind} by ${move.player}`);
    }

    const newHistory: Move[] = [...state.history, move];

    if (move.kind === 'place') {
      const idx = state.players.indexOf(move.player);
      const newHand = state.handAt(idx).without(move.tile);
      const newHands = [...state.hands];
      newHands[idx] = newHand;
      const newChain = state.chain.place(move.tile, move.end);

      const wonByDomino = newHand.count === 0;
      return state.with({
        currentPlayerIndex: wonByDomino ? state.currentPlayerIndex : nextPlayerIndex(state),
        hands: newHands,
        chain: newChain,
        turnNumber: state.turnNumber + 1,
        consecutivePassCount: 0,
        history: newHistory,
        isOver: wonByDomino,
      });
    }

    if (move.kind === 'pass') {
      const newPassCount = state.consecutivePassCount + 1;
      const blocked = newPassCount >= state.players.length;
      return state.with({
        currentPlayerIndex: blocked ? state.currentPlayerIndex : nextPlayerIndex(state),
        turnNumber: state.turnNumber + 1,
        consecutivePassCount: newPassCount,
        history: newHistory,
        isOver: blocked,
      });
    }

    // resign — ends the round immediately; index left unchanged (GetOutcome
    // reads the resigner from history, not CurrentPlayerIndex).
    return state.with({
      turnNumber: state.turnNumber + 1,
      history: newHistory,
      isOver: true,
    });
  }

  getOutcome(state: MatchState): MatchOutcome | null {
    if (!state.isOver) {
      return null;
    }

    const remaining = new Map<PlayerId, number>();
    for (const [i, player] of state.players.entries()) {
      remaining.set(player, state.handAt(i).pipTotal);
    }

    const last = state.history[state.history.length - 1];
    if (last?.kind === 'resign') {
      return this.resignOutcome(state, last.player, remaining);
    }

    // Domino end: someone has zero tiles.
    let dominoWinnerIdx = -1;
    for (let i = 0; i < state.players.length; i++) {
      if (state.handAt(i).count === 0) {
        dominoWinnerIdx = i;
        break;
      }
    }
    if (dominoWinnerIdx >= 0) {
      const winner = state.playerAt(dominoWinnerIdx);
      let score = 0;
      for (let i = 0; i < state.players.length; i++) {
        if (i !== dominoWinnerIdx) {
          score += state.handAt(i).pipTotal;
        }
      }
      return {
        reason: 'domino',
        winnerId: winner,
        winningTeamId: state.partnership.getTeamOf(winner),
        winnerScore: score,
        remainingPips: remaining,
      };
    }

    // Block end: lowest pip total wins; a tie is a scoreless draw.
    return this.blockOutcome(state, remaining);
  }

  private resignOutcome(
    state: MatchState,
    resigner: PlayerId,
    remaining: Map<PlayerId, number>,
  ): MatchOutcome {
    const resignerPips = state.handOf(resigner).pipTotal;
    const reason: MatchEndReason = 'resigned';

    if (state.players.length === 2) {
      const other = state.playerAt(0) === resigner ? state.playerAt(1) : state.playerAt(0);
      return {
        reason,
        winnerId: other,
        winningTeamId: state.partnership.getTeamOf(other),
        winnerScore: resignerPips,
        remainingPips: remaining,
      };
    }

    // 3+ players: lowest pips among the non-resigners wins (same rule as a block).
    let winner: PlayerId | null = null;
    let lowest = Number.MAX_SAFE_INTEGER;
    for (const [i, p] of state.players.entries()) {
      if (p === resigner) {
        continue;
      }
      const pips = state.handAt(i).pipTotal;
      if (pips < lowest) {
        lowest = pips;
        winner = p;
      }
    }
    return {
      reason,
      winnerId: winner,
      winningTeamId: winner !== null ? state.partnership.getTeamOf(winner) : null,
      winnerScore: resignerPips,
      remainingPips: remaining,
    };
  }

  private blockOutcome(state: MatchState, remaining: Map<PlayerId, number>): MatchOutcome {
    let winnerIdx = -1;
    let lowest = Number.MAX_SAFE_INTEGER;
    let tiedAtLowest = false;
    for (let i = 0; i < state.players.length; i++) {
      const pips = state.handAt(i).pipTotal;
      if (pips < lowest) {
        lowest = pips;
        winnerIdx = i;
        tiedAtLowest = false;
      } else if (pips === lowest) {
        tiedAtLowest = true;
      }
    }

    if (tiedAtLowest || winnerIdx < 0) {
      return {
        reason: 'blocked',
        winnerId: null,
        winningTeamId: null,
        winnerScore: 0,
        remainingPips: remaining,
      };
    }

    const winner = state.playerAt(winnerIdx);
    let score = 0;
    for (let i = 0; i < state.players.length; i++) {
      if (i !== winnerIdx) {
        score += state.handAt(i).pipTotal;
      }
    }
    return {
      reason: 'blocked',
      winnerId: winner,
      winningTeamId: state.partnership.getTeamOf(winner),
      winnerScore: score,
      remainingPips: remaining,
    };
  }
}

function nextPlayerIndex(state: MatchState): number {
  return (state.currentPlayerIndex + 1) % state.players.length;
}

function movesEqual(a: Move, b: Move): boolean {
  if (a.player !== b.player) {
    return false;
  }
  if (a.kind === 'place' && b.kind === 'place') {
    return sameTile(a, b) && a.end === b.end;
  }
  return a.kind === 'pass' && b.kind === 'pass';
}

function sameTile(a: PlaceMove, b: PlaceMove): boolean {
  return a.tile.equals(b.tile);
}
